compose_rl/utils/utils.py

```diff
@@ -1237,3 +1252,211 @@ def flatten(coll: Union[Iterable[Any], str]) -> Generator[Any, None, None]:
                 yield subc
         else:
             yield i
+
+
+def _get_params_to_summon_fsdp2(module: torch.nn.Module, recurse: bool = True):
+    """Gets the DTensors to materialize for an FSDP2 model based on recurse.
+
+    If recurse=False, we can encounter the following state:
+    FSDPModule_1
+      |- weight (DTensor)
+      |- FSDPModule_2
+      |   |- weight (DTensor)
+      |- RegularModule_1
+      |   |- weight (DTensor)
+      |   |- FSDPModule_3
+      |   |   |- weight (DTensor)
+    Where summon_full_params(FSDPModule_1) should materialize RegularModule_1.weight
+    alongside the original FSDPModule_1.weight. Therefore, we use a dfs traversal
+    to get all DTensors not owned by downstream FSDPModules.
+    """
+    if recurse:
+        return {
+            name: param
+            for name, param in
+            module.named_parameters(recurse=True, remove_duplicate=False)
+            if isinstance(param, DTensor)
+        }
+
+    dtensor_params = {}
+
+    def _dfs(module: torch.nn.Module, prefix: str = ''):
+        # Add all DTensors within this (FSDP)module
+        for name, param in module.named_parameters(
+            recurse=False,
+            remove_duplicate=False,
+        ):
+            if isinstance(param, DTensor):
+                full_name = f'{prefix}.{name}' if prefix else name
+                dtensor_params[full_name] = param
+        for child_name, child in module.named_children():
+            if isinstance(child, FSDPModule):
+                continue
+            full_name = f'{prefix}.{child_name}' if prefix else child_name
+            _dfs(child, full_name)
+
+    _dfs(module, '')
+    return dtensor_params
+
+
+@contextmanager
+def _summon_full_params_fsdp2(
+    model: torch.nn.Module,
+    writeback: bool = True,
+    recurse: bool = True,
+):
+    """Context manager to get full params for FSDP2 models with DTensor APIs.
+
+    Note: We use DTensor APIs to materialize the full parameters instead of using `unshard`
+    and `reshard` as writeback doesn't seem to work correctly with DTensors
+    that uses DTensor APIs to materialize the full parameters. We currently don't support
+    rank0_only
+    """
+    from torch.distributed.tensor import Replicate, distribute_tensor
+
+    dtensor_params = _get_params_to_summon_fsdp2(model, recurse=recurse)
+
+    if not dtensor_params:
+        yield
+        return
+
+    model_dtensors = {}
+    metadata = {}
+    tied_params = {}
+
+    # We want to get the module and attr of the param, so we can assign
+    # module.attr = param.full_tensor() before we yield and
+    # module.attr = distributed (maybe updated) tensor after we yield.
+    def _get_module_and_attr(model: torch.nn.Module, param_name: str):
```

Comment:
TODO: test this out with `activation_cpu_offloading=True` to make sure there's no issues here